### ISSUE TITLE
Pipenv cannot handle local version identifiers 

### ISSUE THREAD

**Issue by Hakuyume** [pypa/pipenv#5566]
Be sure to check the existing issues (both open and closed!), and make sure you are running the latest version of Pipenv.

Check the [diagnose documentation](https://pipenv.pypa.io/en/latest/diagnose/) for common issues before posting! We may close your issue if it is very similar to one of them. Please be considerate, or be on your way.

Make sure to mention your debugging experience if the documented solution failed.

### Issue description

Pipenv raises a `ParseException` when `Pipfile` contains local version identifiers (e.g. `+abc.5`).
"local version identifiers" is defined in [PEP440](https://peps.python.org/pep-0440/#local-version-identifiers).

### Expected result

Handle local version identifiers correctly

### Actual result

Raise a `ParseException`

### Steps to replicate

```console
$ docker run -it -w /tmp python:3.10 bash                 
                                                           
root@4d892a20b4c0:/tmp# pip install pipenv
...
root@4d892a20b4c0:/tmp# pip freeze
certifi==2022.12.7
distlib==0.3.6
filelock==3.9.0
pipenv==2022.12.19
platformdirs==2.6.2
virtualenv==20.17.1
virtualenv-clone==0.5.7
root@4d892a20b4c0:/tmp# echo 'packages.foo = "1.0+abc.5"' > Pipfile                                                
                                                                                                                   
root@4d892a20b4c0:/tmp# cat Pipfile                                                                                
packages.foo = "1.0+abc.5"
root@4d892a20b4c0:/tmp# pipenv lock --clear                                                                        
Creating a virtualenv for this project...                                                                          
Pipfile: /tmp/Pipfile                                                                                              
Using /usr/local/bin/python3 (3.10.9) to create virtualenv...                                                      
⠋ Creating virtual environment...created virtual environment CPython3.10.9.final.0-64 in 543ms                     
  creator CPython3Posix(dest=/root/.local/share/virtualenvs/tmp-XVr6zr33, clear=False, no_vcs_ignore=False, global=
False)                                                                                                             
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/root/.loc
al/share/virtualenv)                                                                                               
    added seed packages: pip==22.3.1, setuptools==65.6.3, wheel==0.38.4                                            
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator      
                                                                                                                   
✔ Successfully created virtual environment!                                                                        
Virtualenv location: /root/.local/share/virtualenvs/tmp-XVr6zr33                                                   
Locking [packages] dependencies...                                                                                 
Traceback (most recent call last):                                                                                 
  File "/usr/local/lib/python3.10/site-packages/pipenv/patched/pip/_vendor/packaging/requirements.py", line 102, in
 __init__                                                                                                          
    req = REQUIREMENT.parseString(requirement_string)                                                              
  File "/usr/local/lib/python3.10/site-packages/pipenv/patched/pip/_vendor/pyparsing/core.py", line 1141, in parse_
string                                                                                                             
    raise exc.with_traceback(None)                       
pipenv.patched.pip._vendor.pyparsing.exceptions.ParseException: Expected string_end, found '+'  (at char 6), (line:
1, col:7)

During handling of the above exception, another exception occurred:                                                

Traceback (most recent call last):                       
  File "/usr/local/lib/python3.10/site-packages/pipenv/patched/pip/_vendor/pkg_resources/__init__.py", line 3101, i
n __init__
    super(Requirement, self).__init__(requirement_string)                                                          
  File "/usr/local/lib/python3.10/site-packages/pipenv/patched/pip/_vendor/packaging/requirements.py", line 104, in
 __init__
    raise InvalidRequirement(                            
pipenv.patched.pip._vendor.packaging.requirements.InvalidRequirement: Parse error at "'+abc.5'": Expected string_en
d

During handling of the above exception, another exception occurred:
Traceback (most recent call last):                       
  File "/usr/local/bin/pipenv", line 8, in <module>                                                                
    sys.exit(cli())                                      
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)                    
  File "/usr/local/lib/python3.10/site-packages/pipenv/cli/options.py", line 57, in main
    return super().main(*args, **kwargs, windows_expand_args=False)                                                
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/click/core.py", line 1053, in main
    rv = self.invoke(ctx)                                
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))                                                        
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)                                                                 
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)                   
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/click/decorators.py", line 84, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)                                                                     
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)                   
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)                                                               
  File "/usr/local/lib/python3.10/site-packages/pipenv/cli/command.py", line 340, in lock
    do_lock(                                             
  File "/usr/local/lib/python3.10/site-packages/pipenv/core.py", line 1190, in do_lock
    venv_resolve_deps(                                   
  File "/usr/local/lib/python3.10/site-packages/pipenv/utils/resolver.py", line 1035, in venv_resolve_deps
    deps = convert_deps_to_pip(deps, project, include_index=True)                                                  
  File "/usr/local/lib/python3.10/site-packages/pipenv/utils/dependencies.py", line 283, in convert_deps_to_pip
    new_dep = Requirement.from_pipfile(dep_name, dep)                                                              
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/requirementslib/models/requirements.py", line 2750, i
n from_pipfile
    r = NamedRequirement.from_pipfile(name, pipfile)                                                               
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/requirementslib/models/requirements.py", line 1371, i
n from_pipfile
    req = init_requirement("{0}{1}".format(name, version))                                                         
  File "/usr/local/lib/python3.10/site-packages/pipenv/vendor/requirementslib/models/utils.py", line 191, in init_r
equirement
    req = Requirement.parse(name)                        
  File "/usr/local/lib/python3.10/site-packages/pipenv/patched/pip/_vendor/pkg_resources/__init__.py", line 3148, i
n parse
    req, = parse_requirements(s)                         
  File "/usr/local/lib/python3.10/site-packages/pipenv/patched/pip/_vendor/pkg_resources/__init__.py", line 3094, i
n parse_requirements
    yield Requirement(line)                              
  File "/usr/local/lib/python3.10/site-packages/pipenv/patched/pip/_vendor/pkg_resources/__init__.py", line 3103, i
n __init__
    raise RequirementParseError(str(e))                  
pipenv.patched.pip._vendor.pkg_resources.RequirementParseError: Parse error at "'+abc.5'": Expected string_end
```
-------------------------------------------------------------------------------
<details><summary>$ pipenv --support</summary>

Pipenv version: `'2022.12.19'`

Pipenv location: `'/usr/local/lib/python3.10/site-packages/pipenv'`

Python location: `'/usr/local/bin/python'`

OS Name: `'posix'`

User pip version: `'22.3.1'`

user Python installations found:

  - `3.10.9`: `/usr/local/bin/python3`
  - `3.10.9`: `/usr/local/bin/python`
  - `3.9.2`: `/usr/bin/python3`
  - `3.9.2`: `/usr/bin/python3.9`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.10.9',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '5.18.13-arch1-1',
 'platform_system': 'Linux',
 'platform_version': '#1 SMP PREEMPT_DYNAMIC Fri, 22 Jul 2022 13:05:04 +0000',
 'python_full_version': '3.10.9',
 'python_version': '3.10',
 'sys_platform': 'linux'}
```

System environment variables:

  - `HOSTNAME`
  - `PYTHON_VERSION`
  - `PWD`
  - `PYTHON_SETUPTOOLS_VERSION`
  - `HOME`
  - `LANG`
  - `GPG_KEY`
  - `TERM`
  - `SHLVL`
  - `PYTHON_PIP_VERSION`
  - `PYTHON_GET_PIP_SHA256`
  - `PYTHON_GET_PIP_URL`
  - `PATH`
  - `_`
  - `PIP_DISABLE_PIP_VERSION_CHECK`
  - `PIP_PYTHON_PATH`
  - `PYTHONDONTWRITEBYTECODE`
  - `PYTHONFINDER_IGNORE_UNSUPPORTED`

Pipenv–specific environment variables:

Debug–specific environment variables:

  - `PATH`: `/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`
  - `LANG`: `C.UTF-8`
  - `PWD`: `/tmp`

---------------------------

Contents of `Pipfile` ('/tmp/Pipfile'):

```toml
packages.foo = "1.0+abc.5"

```

</details>

**Comment by Hakuyume** [comment]
If I modify `PUNCTUATION` in `pipenv/patched/pip/_vendor/packaging/requirements.py` to include `+` as punctuation character, it works as expected.

**Comment by matteius** [comment]
That is a bundled version of the latest pip ... Perhaps then this is an issue report for pip?  Does regular pip work for this use case?

**Comment by Hakuyume** [comment]
> Does regular pip work for this use case?

With regular pip, the parsing phase works (installation fails since `foo` is not published to pypi).

```console
$ docker run -it -w /tmp python:3.10 bash
root@88a3883f5d14:/tmp# echo 'foo==1.0+abc.5' > requirements.txt
root@88a3883f5d14:/tmp# pip install -r requirements.txt 
ERROR: Could not find a version that satisfies the requirement foo==1.0+abc.5 (from versions: none)
ERROR: No matching distribution found for foo==1.0+abc.5
```

**Comment by Hakuyume** [comment]
Regular pip seems to accept invalid version specifications...
Testing using `pip install -r` may be wrong.
```console
root@88a3883f5d14:/tmp# echo 'foo==a+b-c*d/e^f' > requirements.txt
root@88a3883f5d14:/tmp# pip install -r requirements.txt 
ERROR: Could not find a version that satisfies the requirement foo==a+b-c*d/e^f (from versions: none)
ERROR: No matching distribution found for foo==a+b-c*d/e^f
```

**Comment by matteius** [comment]
I believe this has been solved for, `2023.9.1`

```
matte@LAPTOP-N5VSGIBD MINGW64 ~/Projects/pipenv_editable_install_repro
$ pipenv sync
Loading .env environment variables...
Creating a virtualenv for this project...
Pipfile: C:\Users\matte\Projects\pipenv_editable_install_repro\Pipfile
Using c:/users/matte/.pyenv/pyenv-win/versions/3.9.13/python3.exe (3.9.13) to create virtualenv...
[ ===] Creating virtual environment...created virtual environment CPython3.9.13.final.0-64 in 1817ms
  creator CPython3Windows(dest=C:\c\users\matte\.virtualenvs\pipenv_editable_install_repro-crT5IkmP, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=C:\Users\matte\AppData\Local\pypa\virtualenv)
    added seed packages: pip==23.2.1, setuptools==68.0.0, wheel==0.41.1
  activators BashActivator,BatchActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator

Successfully created virtual environment!
Virtualenv location: C:\c\Users\matte\.virtualenvs\pipenv_editable_install_repro-crT5IkmP
Installing dependencies from Pipfile.lock (d55127)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
All dependencies are now up-to-date!

matte@LAPTOP-N5VSGIBD MINGW64 ~/Projects/pipenv_editable_install_repro
$ pipenv run pip freeze
Loading .env environment variables...
# Editable install with no version control (test-hello==1.0+abc.5)
-e c:\users\matte\projects\pipenv_editable_install_repro\applications\test-hello
```
